### ISSUE TITLE
Support `waitid(P_PIDFD, ...)`

### ISSUE THREAD

**Issue by rocallahan** [mozilla/rr#2666]
I landed a test for this in ad9451f858cf518c07de05183849d0f8374b90e3 but I don't know how to implement it.

If the pidfd refers to a tracee task, then we need to get the pid of that task and process it like `P_PID` ... i.e. possibly emulating the wait.

If the pidfd refers to a non-tracee task, then no emulation is needed and we can just pass the syscall through.

The problem is, I don't know how to get the pid from a pidfd :-(.

**Comment by rocallahan** [comment]
Ah, `/proc/.../fdinfo/...` has it.

**Comment by rocallahan** [comment]
https://github.com/mozilla/rr/commit/7ae1e53c9db1bc4eafa4dcf1cf5f76ac8670b4bb fixed this. However, there are still some potential issues with pidfds. In particular, poll() etc might require an emulated wait, which we don't support, and I'm not sure how we would support it :-(.

**Comment by brauner** [comment]
Yep, the `fdinfo` file has that info. :) You can also see ways how to parse this in the selftests I added here:
https://github.com/torvalds/linux/blob/fc4f28bb3daf3265d6bc5f73b497306985bb23ab/tools/testing/selftests/pidfd/pidfd_open_test.c#L87
or systemd has logic for this as well:
https://github.com/systemd/systemd/blob/d7b34e384108ff431401d942ec0daaf156ec27c1/src/basic/process-util.c#L1535

**Comment by brauner** [comment]
Oh, what is the issue with `poll()` you're referring too?

**Comment by brauner** [comment]
I'm the maintainer of this stuff in the kernel so feel free to mention stuff that you need or could be improved. :)

**Comment by rocallahan** [comment]
There are a couple of cases where rr has to emulate syscalls that wait on a task.

One case involves ptrace(). To support record and replay of tasks that ptrace() other tasks, rr has to emulate ptrace() for those tasks, since rr ptraces everything it records, and a task can't have multiple ptracers. When a task T1 waits for T2 to get ptrace events, we need to emulate the waiting and delivery of those ptrace stops because T1 is not really ptracing T2. Also, when T1 waits for a task T2 that it thinks it is ptracing, but T2 is not a child of T1, the kernel wouldn't even deliver exit events to T1, so we have to emulate in this case too.

Another case involves SIGSTOP/SIGCONT. When a task T2 we're recording receives SIGSTOP we have to emulate it stopping (delivering the SIGSTOP to T2 for real interferes with rr's interactions with T2; I forget the details right now), and likewise we need to emulate SIGCONT. Waits with `WUNTRACED` should observe these stops so we need to emulate those waits too.

So the question is, how does this emulation interact with pidfds? The concern is if a task T1 poll()s a pidfd referring to a task T2 for which we do an emulated stop, and that emulated stop needs to change the state of the pidfd to wake up T1. From the `pidfd_open` man page, it appears that only actual task exit can currently wake the pidfd, in which case I think there isn't a problem for us since although we sometimes delay task exits we never do so indefinitely. But if, currently or in the future, pidfds can be notified for non-exit task status changes such as ptrace events, then we would have a problem.
